client/lib/domains/dns/index.js

```diff
@@ -86,32 +75,44 @@ function isValidData( data, type ) {
 			return data.match( /^[a-f0-9\:]+$/i );
 		case 'CNAME':
 		case 'MX':
-			return isValidDomainName( data, type );
+			return isValidDomain( data );
 		case 'TXT':
 			return data.length < 256;
 	}
 }
 
-function getNormalizedData( fieldValues, selectedDomainName ) {
-	var data = fieldValues;
-
-	data.data = getFieldWithDot( data.data );
-
-	if ( includes( [ 'A', 'AAAA' ], data.type ) ) {
-		data.name = removeTrailingDomain( data.name, selectedDomainName );
+function getNormalizedData( record, selectedDomainName ) {
+	const normalizedRecord = Object.assign( {}, record );
+	normalizedRecord.data = getFieldWithDot( record.data );
+	normalizedRecord.name = getNormalizedName( record.name, record.type, selectedDomainName );
+	if ( record.target ) {
+		normalizedRecord.target = getFieldWithDot( record.target );
 	}
 
-	data.name = getFieldWithDot( data.name );
+	return normalizedRecord;
+}
+
+function getNormalizedName( name, type, selectedDomainName ) {
+	const isSrvWithEmptyName = 'SRV' === type && ! name,
+		endsWithDomain = endsWith( name, '.' + selectedDomainName );
 
-	if ( data.target ) {
-		data.target = getFieldWithDot( data.target );
+	if ( isSrvWithEmptyName ) {
+		return name;
 	}
 
-	return data;
+	if ( endsWithDomain ) {
+		if ( isARecord( type ) ) {
+			return name.replace( new RegExp( '\\.+' + selectedDomainName + '\\.?$', 'i' ), '' );
+		}
+		return getFieldWithDot( name );
+	} else if ( ! isARecord( type ) ) {
+		return name + '.' + selectedDomainName + '.';
+	}
+	return name;
 }
 
-function removeTrailingDomain( domain, trailing ) {
-	return domain.replace( new RegExp( '\\.+' + trailing + '\\.?$', 'i' ), '' );
+function isARecord( type ) {
+	return includes( [ 'A', 'AAAA' ], type );
```

Comment:
Heh, it was a mental shortcut, but you are right, the function's name could be misleading :+1: